src/protocols/bdx/AsyncTransferFacilitator.h

```diff
@@ -0,0 +1,136 @@
+
+    /**
+     * This method should be implemented to destroy the object subclassing AsyncTransferFacilitator.
+     */
+    virtual void DestroySelf() = 0;
+
+    void ProcessOutputEvents();
+
+    void CleanUp();
+
+    // The transfer session coresponding to this AsyncTransferFacilitator object.
+    TransferSession mTransfer;
+
+private:
+    bool mProcessingOutputEvents = false;
+
+    // The Exchange holder that holds the exchange context used for sending and receiving BDX messages.
+    Messaging::ExchangeHolder mExchange;
+
+    // The timeout for the BDX transfer session.
+    System::Clock::Timeout mTimeout;
+
+    System::Layer * mSystemLayer;
+};
+
+/**
+ * An AsyncTransferFacilitator that is initialized to respond to an incoming BDX transfer request.
+ * An AsyncResponder object is associated with an exchange and handles all BDX messages sent over that exchange.
+ *
+ * Provides a method for initializing the TransferSession members but still needs to be extended to implement
+ * HandleTransferSessionOutput.
+ *
+ * An instance of some subclass of this class should be used as the exchange delegate for a BDX transfer.
+ */
+class AsyncResponder : public AsyncTransferFacilitator
+{
+public:
+    /**
+     * Initialize the TransferSession state machine to be ready for an incoming transfer request.
+     *
+     * @param[in] exchangeCtx     The exchange to use for the transfer.
+     * @param[in] role            The role of the Responder: Sender or Receiver of BDX data
+     * @param[in] xferControlOpts Supported transfer modes (see TransferControlFlags)
+     * @param[in] maxBlockSize    The maximum supported size of BDX Block data
+     * @param[in] timeout         The chosen timeout delay for the BDX transfer
+     */
+    CHIP_ERROR Init(System::Layer * layer, Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                    BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize, System::Clock::Timeout timeout);
+
+    /**
+     * This is called by the subclass implementing HandleTransferSessionOutput to notify the AsyncResponder
+     * that it has handled the OutputEvent specified in "event" and "status" is the result of handling the event.
+     * If the OutputEvent was handled successfully by the subclass that means the TransferSession API call to handle the output
+     * event was successful and the TransferSession has the next message prepared. In that case, the "status" is set to
+     * CHIP_NO_ERROR and the AsyncResponder should call ProcessOutputEvents to get and process the next output events from the
+     * TransferSession to proceed with the BDX transfer. Otherwise if the output event indicates the end of BDX transfer (like EOF
+     * receieved or transfer timeout), it will clean up and destroy itself and the subclass. If the "status" is set to an error
+     * code, the AsyncResponder should abort the transfer and clean up once the status report is sent over the exchange.
```

Comment:
Looks like https://github.com/project-chip/connectedhomeip/pull/29364#discussion_r1777745598 was not addressed and this is still talking about implementation details, not things about how the API must/should be used.  So how about the following, but _please_ carefully check that the claims I am making in this comment are in fact correct.  Once we are agreed on the API behavior here, I'll check that the implementation matches.
```suggestion
     * Method that must be called by the subclass implementing HandleTransferSessionOutput to notify the AsyncResponder
     * that it has handled the OutputEvent specified in "event" and "status" is the result of handling the event.
     *
     * Every call to HandleTransferSessionOutput must result in a call to NotifyEventHandled.  The call
     * to NotifyEventHandled may happen before HandleTransferSessionOutput returns, or may happen
     * later, asynchronously.
```

Specific questions about that text:

1. Is it true that all `HandleTransferSessionOutput` must result in `NotifyEventHandled`?  Or can they result in the subclass doing `CleanUp()` directly (e.g. when stack shutdown is happening)?
2. Is it true that NotifyEventHandled can happen either sync or async?
3. None of the discussion about ProcessOutputEvents is relevant here; that's implementation details that consumers of this API don't care about and should not care about.
4. The discussion about "clean up and destroy itself" is not relevant here, since that happens async, right?  So there is nothing special the API consumer has to worry about here with regard to that?
5. That said, we are queueing an async thing on the system layer.  If we are in shutdown and call `CleanUp`, will that async thing get run before system layer is shut down?  That is, does system layer shutdown process pending ScheduleWork bits, or do we leak in this situation?

Reply:
1. True
2. True
3. Ok.
4. Ok.
5. CleanUp should only be called from NotifyEventHandled or Sendmessage if message failed to send. I am calling it from onExchangeClosing/onResponseTimeout. For the latter 2 cases we need to evaluate if calling cleanup is the right thing to do or we can just release the exchange.